crates/polars-plan/src/dsl/function_expr/range/int_range.rs

```diff
@@ -1,29 +1,70 @@
 use polars_core::prelude::*;
 use polars_core::series::{IsSorted, Series};
+use polars_core::with_match_physical_integer_polars_type;
 
 use super::utils::ensure_range_bounds_contain_exactly_one_value;
 
-pub(super) fn int_range(s: &[Series], step: i64) -> PolarsResult<Series> {
-    let start = &s[0];
-    let end = &s[1];
+pub(super) fn int_range(s: &[Series], step: i64, dtype: DataType) -> PolarsResult<Series> {
+    let mut start = &s[0];
+    let mut end = &s[1];
 
     ensure_range_bounds_contain_exactly_one_value(start, end)?;
 
-    match start.dtype() {
-        dt if dt == &IDX_DTYPE => {
-            let start = start.idx()?.get(0).unwrap();
-            let end = end.cast(&IDX_DTYPE)?;
-            let end = end.idx()?.get(0).unwrap();
-            int_range_impl::<IdxType>(start, end, step)
-        },
-        _ => {
-            let start = start.cast(&DataType::Int64)?;
-            let end = end.cast(&DataType::Int64)?;
-            let start = start.i64()?.get(0).unwrap();
-            let end = end.i64()?.get(0).unwrap();
-            int_range_impl::<Int64Type>(start, end, step)
-        },
+    let (start_storage, end_storage);
+    if *start.dtype() != dtype {
+        start_storage = start.strict_cast(&dtype)?;
+        start = &start_storage;
     }
+    if *end.dtype() != dtype {
+        end_storage = end.strict_cast(&dtype)?;
+        end = &end_storage;
+    }
+
+    with_match_physical_integer_polars_type!(dtype, |$T| {
+        let start_v = get_first_series_value::<$T>(start)?;
+        let end_v = get_first_series_value::<$T>(end)?;
+        int_range_impl::<$T>(start_v, end_v, step)
+    })
+}
+
+fn get_first_series_value<T>(s: &Series) -> PolarsResult<T::Native>
+where
+    T: PolarsIntegerType,
+{
+    let ca: &ChunkedArray<T> = s.as_any().downcast_ref().unwrap();
+    let value_opt = unsafe { ca.get_unchecked(0) };
```

Comment:
Getting a single value is not worth using `unsafe`. We also didn't seem to have catched the empty case, so this would segfault if I am not mistaken.